src/transformers/models/mega/modeling_mega.py

```diff
@@ -0,0 +1,2299 @@
+                if padding_mask is not None:
+                    padding_mask = padding_mask.unsqueeze(1)
+            else:
+                # B x L x S -> B x K x C x S
+                nc = ctx_len // self.config.chunk_size
+                k = k.reshape(bsz, nc, self.config.chunk_size, self.config.shared_representation_size)
+                v = v.reshape(bsz, nc, self.config.chunk_size, self.config.intermediate_size)
+                if padding_mask is not None:
+                    padding_mask = padding_mask.view(bsz, nc, self.config.chunk_size)
+
+        # this is in the original Mega implementation to work around fork/join parallelism not supporting optional types
+        if padding_mask is not None and padding_mask.dim() == 0:
+            padding_mask = None
+
+        attn_weights = self.attention_function(q, k, padding_mask=padding_mask, causal_mask=causal_mask)
+
+        v = self.hidden_dropout(v, batch_first=True)
+        kernel = self.attention_dropout(attn_weights)
+
+        # B x K x C x E -> B x L x E -> L x B x E
+        h = torch.matmul(kernel, v).view(bsz, seq_len, self.config.intermediate_size).transpose(0, 1)
+
+        # L x B x E -> L x B x D
+        h = self.activation(hx + self.h_proj(h * r))
+        h = self.dropout(h)
+        # L x B x D
+        out = torch.addcmul(residual, u, h - residual)
+
+        if not self.config.normalize_before_mega:
+            out = self.norm(out)
+
+        return_values = (out, attn_weights) if output_attentions else (out,)
+
+        if self.config.is_decoder:
+            return_values = return_values + (updated_self_key, updated_self_value, updated_ema_state)
+
+        return return_values
+
+
+# Normalized feed-forward network
+# left as-is from original Mega repo aside from retrieving args from Hugging Face config
+class NormalizedFeedForwardNetwork(nn.Module):
+    """
+    Normalized feed-forward network used in Mega blocks
+    """
+
+    def __init__(self, config: MegaConfig):
+        super().__init__()
+
+        self.config = config
+        self.hidden_dim = config.nffn_hidden_size
+        self.act_fn = config.activation
+        self.activation = get_activation_fn(config.activation)
+
+        dropout_module = MegaFeatureDropout if self.config.use_feature_dropout else MegaDropout
+        self.dropout = dropout_module(self.config.dropout_prob, module_name=self.__class__.__name__)
+        self.hidden_dropout = dropout_module(
+            self.config.nffn_activation_dropout_prob, module_name=self.__class__.__name__
+        )
+
+        self.prenorm = self.config.normalize_before_ffn
+        self.norm = SequenceNorm(
+            self.config.normalization_type, self.config.hidden_size, affine=self.config.norm_affine
+        )
+
+        self.fc1 = nn.Linear(self.config.hidden_size, self.config.nffn_hidden_size)
+        self.fc2 = nn.Linear(self.config.nffn_hidden_size, self.config.hidden_size)
+
+        self.reset_parameters()
```

Comment:
unless this is specific to MEGA as you would need to run this during inference and training, should not be needed

Reply:
noted!